New native hosting

vitek-karas opened the pull request:
Very first draft of the general thinking on the new native hosting API (the advanced scenario).

Please don't share yet.

vitek-karas committed:
First draft

vitek-karas commented:
@elinor-fung

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,182 @@
+
+### Goals
+* All hosts should be able to use the new API (whether they will is a separate question as the old API has to be kept for backward compat reasons)
+* Try to hide as much implementation details as possible
```
>Try to hide as much implementation details as possible

Hide implementation details as much as possible

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,182 @@
+* Try to hide as much implementation details as possible
+  * Makes the API generally easier to use/understand
+  * Gives the implementation more freedom
+  * Should allow future improvements without breaking the API
```
~Makes~ -> Make
~Gives~ -> Give
~Should allow~ -> Allow

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,182 @@
+* Extensible
+
+## New scenarios
+The API should allow these, doesn't have to support them all in the beginning. It should be possible to ship with only some of these supported.
```
I'd move the comments about not supporting everything at first to after the list of scenarios. Something like:
```
The API should allow these scenarios:
<...>
It should be possible to ship with only some of these supported, then enable more scenarios afterwards.
```

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,182 @@
+* Runtime properties
+  * Ability to specify additional runtime properties from the native host
+  * Ability for the host to implement conflict resolution for runtime properties
+  * Ability for the host to inspect calculated runtime properties (the ones calculated by `hostfxr`/`hostpolicy`)
```
Having 'ability' so many times seems a bit redundant. Maybe something like:
* Runtime properties - the native host has the ability to:
  * Specify...
  * Implement...
  * Inspect...

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,182 @@
+struct hostfxr_initialize_parameters
+{
+    size_t size;
+    const char_t * host_path;   // path to the .exe
```
.exe -> native host ?

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,182 @@
+
+The `hostfxr_initialize_parameters` structure stores initialize parameters which are common to all forms of initialization.
+* `size` - the size of the structure. This is used for versioning. Should be set to `sizeof(hostfxr_initialize_parameters)`.
+* `host_path` - path to the native host (typically the `.exe`). This value is not used for anything by the host, it's just passed to the CoreCLR as the path to the executable. It can point to a file which is not executable itself, if such file doesn't exist (for example in COM activation scenarios this points to the `comhost.dll`).
```
'the host' is a bit overloaded in this world - native host, (initialized) host, the host. It might be worth going through and making sure each usage is explicit - and maybe adding a short definition / note on terminology at the top.

vitek-karas replied:
Tried to do what you suggested - hopefully it's clearer now.

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
````diff
@@ -0,0 +1,182 @@
+);
+```
+
+Initializes the host for running a managed application. Also used for running SDK commands.
````
~. Also~ -> - also

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,182 @@
+
+This function can only be called once per-process. It's not supported to run multiple apps in one process (even sequentially).
+
+*Note: This is effectively a replacement for the `hostfxr_main_startupinfo` and `hostfxr_main` - both of those APIs should be possible to implement using the `hostfxr_initialize_for_app` and the `hostfxr_run_app`.*
```
Remove 'the' before the function names

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,182 @@
+* Find the root framework (`Microsoft.NETCore.App`) and load the `hostpolicy` from it
+* The `hostpolicy` will then process all relevant `.deps.json` files and produce the TPA, native search paths and other artifacts.
+
+The functions will NOT load the CoreCLR runtime, they just prepare everything to the point where it can be loaded.
```
~, they~ -> . They

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,182 @@
+The functions will NOT load the CoreCLR runtime, they just prepare everything to the point where it can be loaded.
+
+
+### Initialized host inspection and modification
```
Inspect and modify the initialized host

(Or change the other headings to be nouns instead of actions if you prefer it that way. I just want them consistent.)

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
````diff
@@ -0,0 +1,182 @@
+``` C++
+using hostfxr_string_result_fn = void(*) (const char_t * value);
+
+int hostfxr_get_runtime_property(const hostfxr_handle handle, const char_t * name, hostfxr_string_result_fn value_result);
````
I assume that checking whether or not a property is set would be through checking if `hostfxr_get_runtime_property` gives `nullptr` as the value?

vitek-karas replied:
I changed the API - it will probably return a specific error code in the case of missing property.

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,182 @@
+* `handle` - the initialized host handle.
+* `name` - the name of the runtime property to get/set.
+* `value_result` - callback which is called with the value of the property.  
+*TODO: Is this the right API - maybe it should use the "buffer, buffer size" approach where the caller supplies allocated memory instead*
```
Can there be duplicates at this point? If the config has duplicates, the last wins, right? But if config properties collide with something hostpolicy adds, are the duplicates kept?

vitek-karas replied:
Unfortunately there can be duplicates... have not yet figure out how to solve this.
Current runtime behavior is to fail early during startup if there are duplicates, so it's a situation which should really not happen, but we should have a defined behavior. Ideas?

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,182 @@
+
+*TODO: Should there be a way to enumerate all the properties?*
+
+*Maybe for the first release we don't support `hostfxr_get_runtime_property` as that requires the initialization to run all the way through to `hostpolicy` to calculate all properties correctly.*
```
This would mean not supporting conflict resolution at first, right? I thought that was one of the things people were pushing for on the other doc?

If the native host can't inspect before setting and sets values before hostpolicy figures out all the properties, won't that have the same potential for duplicates or confusion around what 'wins'?

vitek-karas replied:
Mostly correct. The goal of conflict resolution is not to resolve conflicts between `.runtimeconfig.json` and the hosting componentes - at least not now.
The goal is to be able to resolve "conflicts" between the app (so `.runtimeconfig.json` mostly) and the native host which may want to set some properties on its own.

The sample from ASP.NET is initial stack size. They want to specify some explicit default from the native host, but if the app defines its own value, they want to somehow reconcile (not just overwrite what the app has).

We should try to make the getter work... if it's too hard, we'll revisit this.

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,182 @@
+* `parameters` - additional parameters - see `hostfxr_initialize_parameters` for details. (Could be made optional potentially)
+* `handle` - output parameter. On success receives an opaque value which identifies the initialized host. The handle should be closed by calling `hostfxr_shutdown`.
+
+This function can only be called once per-process. It's not supported to run multiple apps in one process (even sequentially).
```
Can it be mixed with `hostfxr_initialize_for_runtime_config`?
Does this function fail if the runtime is already loaded through some other mechanism (i.e. not from a previous `hostfxr_initialize_for_app` and `hostfxr_run_app`?

vitek-karas replied:
It can be mixed - one app many components (runtime configs).
Yes - the function will fail if there already is runtime in the process. Running an app "takes over" the process in many ways - this is runtime behavior and we don't have a reason to change it now.

vitek-karas replied:
I added statements to this effect in the document.

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,182 @@
+
+This function can be called multiple times in a process.
+* If it's called when no runtime is present, it will run through the steps to "initialize" the runtime (so resolving frameworks and so on).
+* If it's called when there already is CoreCLR in the process (loaded through the `hostfxr`, direct usage of `coreclr` is not supported), then the function determines if the specified runtime configuration is compatible with the existing runtime and frameworks. If it is, it returns a valid handle, otherwise it fails.
```
Does this mean that if `hostfxr_initialize_for_runtime_config` is called with configs that are not compatible, but the runtime is not actually started yet, the expectation is that the initialize calls will succeed / give valid handles and the failure will come when trying to start the runtime (after the first call)?

vitek-karas replied:
That's an interesting point - I think the answer is no - one of the initialize calls would fail. Due to some limitations we can't jsut look for CoreCLR in the process, instead we rely on global variable in the hosting components. This is something we may need to implement - I know that `comhost` introduces some parts of this, but we may need to extend it even more broadly.

But the general point is, `hostfxr` will have a way to track the "loaded runtime". We could make it so that initializing the host == marking the runtime as loaded (even though we haven't really loaded CoreCLR yet, but we did load `hostpolicy` which is effectively part of the runtime).

elinor-fung replied:
I think failing at the initialize stage makes sense and would be what I want as the caller. Maybe we should clarify that having the runtime present / in the process means is not the same as and does not require starting the runtime (as in the heading in the later section)? Or maybe the pivot is more like 'hosting components have not been initialized' or 'runtime has not been initialized'?

vitek-karas replied:
I added a detailed description - we will have to look into this some more during the implementation. COM activation effectively needs to be able to initialize/run from multiple threads at the same time - so we need to have a solid mechanism in place.

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,182 @@
+* `handle` - handle to the initialized host.
+* `type` - the type of runtime functionality requested *TODO - exact names*
+  * `load_assembly` - entry point which loads an assembly (with dependencies) and returns function pointer for a specified static method.
+  * `com` - COM activation entry-point
```
Maybe link to the COM activation doc?

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,182 @@
+  * `ijw` - IJW entry-point
+* `delegate` - when successful, the native function pointer to the requested runtime functionality.
+
+*TODO: Is it OK to allow calling this even when initialized via `hostfxr_initialize_for_app`? It would mean the app can't be executed via `hostfxr_run_app` anymore.*
```
Is this because `hostfxr_run_app` fails if the runtime is already loaded?

vitek-karas replied:
No - not really. Technically I can't think of a reason to not allow this, other than we don't have a scenario right now. I modified the wording to that effect.

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,182 @@
+* `handle` - handle to the initialized host to close.
+
+
+Notes:
```
Make this a heading? I interpreted this as notes only about `hostfxr_shutdown`

vitek-karas replied:
Completely removed.

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,182 @@
+
+
+Notes:
+* `hostfxr_handle` is currently not absolutely necessary mainly because it won't be supported to call initialize multiple times. The limitation comes from usage of static variables in the implementation (and lack of scenarios). In the future this limitation may go away. Once it does the handle will be necessary for property functionality.
```
>won't be supported to call initialize multiple times

I thought `hostfxr_initialize_for_runtime_config` could be called multiple times?

>property functionality

I'm not sure what this is referring to? Functionality to set/get runtime properties?

vitek-karas replied:
I was mostly wrong - completely removed this.

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,182 @@
+    &params,
+    &handle);
+
+hostfxr_set_runtime_property(handle, "TEST_PROPERTY", "TRUE");
```
Maybe include inspecting / getting in the sample as well?

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,182 @@
+# Native hosting with explicit initialize
+
+### Goals
+* All hosts should be able to use the new API (whether they will is a separate question as the old API has to be kept for backward compat reasons)
```
Do we have a link (or doc) for 'the old API'

vitek-karas replied:
Unfortunately no - it's a big hole... we need to fill it eventually.

vitek-karas committed:
Improvements

Remove some TODOs, clarify some things, better descriptions.

vitek-karas committed:
PR feedback

vitek-karas commented:
Hopefully reacted to all feedback - lot of changes. Please give it another read.

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,209 @@
+
+The `hostfxr_initialize_parameters` structure stores parameters which are common to all forms of initialization.
+* `size` - the size of the structure. This is used for versioning. Should be set to `sizeof(hostfxr_initialize_parameters)`.
+* `host_path` - path to the native host (typically the `.exe`). This value is not used for anything by the hosting components, it's just passed to the CoreCLR as the path to the executable. It can point to a file which is not executable itself, if such file doesn't exist (for example in COM activation scenarios this points to the `comhost.dll`).
```
>components, it's

components. It's

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,209 @@
+
+The functions return a handle to a new host context:
+* The handle must be closed via `hostfxr_shutdown`.
+* The handle is not thread safe - it is only allowed to call functions on it from one thread at a time.
```
>it is only allowed to call

the consumer should only call

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,209 @@
+int hostfxr_initialize_for_app(
+    int argc,
+    const char_t * argv[],
+    const char_t * app_path  
```
```suggestion
    const char_t * app_path,  
```

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,209 @@
+
+This function will fail if there already is a CoreCLR running in the process.
+
+*Note: This is effectively a replacement for `hostfxr_main_startupinfo` and `hostfxr_main`. Currently it should not be a goal to fully replace these APIs because they also support SDK commands which are special in lot of ways and don't fit well with the rest of the native hosting. There's also really no scenario right now which would require the ability to issue SDK commands from a native host. T ...
```
>should not be a goal

is not a goal

>also really no scenario

no scenario

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
````diff
@@ -0,0 +1,209 @@
+
+``` C++
+int hostfxr_initialize_for_runtime_config(
+    const char_t * runtime_config_path
````
```suggestion
    const char_t * runtime_config_path,
```

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
````diff
@@ -0,0 +1,209 @@
+);
+```
+
+This function would load the specified `.runtimeconfig.json`, resolve all frameworks and resolve all the assets from those frameworks. Then prepare runtime initialization where the TPA contains only frameworks. Note that this case does NOT consume any `.deps.json` from the app/component (only processes the framework's `.deps.json`). This entry point is intended for `comhost`/`ijwhost`/`nethost` a ...
````
>frameworks and resolve all the assets from those frameworks. Then 

frameworks, resolve all the assets from those frameworks, and then

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,209 @@
+* If it's called when no runtime is present, it will run through the steps to "initialize" the runtime (resolving frameworks and so on).
+* If it's called when there already is CoreCLR in the process (loaded through the `hostfxr`, direct usage of `coreclr` is not supported), then the function determines if the specified runtime configuration is compatible with the existing runtime and frameworks. If it is, it returns a valid handle, otherwise it fails.
+
+It needs to be possible to call this function simultaneously from multiple threads at the same time as well.
```
~as well~

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,209 @@
+* `value_buffer_size` - the size of the `value_buffer` is `char_t` units.
+* `value_buffer_used` - when successful contains number of `char_t` units used in the buffer (including the null terminator). If the specified `value_buffer_size` was too small, the function returns `HostApiBufferTooSmall` and sets the `value_buffer_used` to the minimum required size.
+
+Trying to get a property which doesn't exist is an error and `hostfxr_get_runtime_property` will return appropriate error code.
```
appropriate -> an appropriate

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
````diff
@@ -0,0 +1,209 @@
+    const char_t * value);
+```
+
+These functions allow the native host to inspect and modify runtime properties. After initialization the properties should contain everything which the hosting components calculates. This gives the native host the opportunity to inspect, modify and resolve any potential conflicts.
````
calculates -> calculate

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,209 @@
+* `name` - the name of the runtime property to get/set. Must not be `nullptr`.
+* `value` - the value of the property to set. When set to `nullptr` the property is "unset" - removed from the runtime property collection.
+* `value_buffer` - buffer to receive the value of the property
+* `value_buffer_size` - the size of the `value_buffer` is `char_t` units.
```
~is~ -> in

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,209 @@
+
+Trying to get a property which doesn't exist is an error and `hostfxr_get_runtime_property` will return appropriate error code.
+
+Setting properties is only supported when the host context was initialized without any runtime in the process. This is really a limitation of the runtime for which the runtime properties are immutable.
```
>without any runtime in the process

Kind of related to the thread above about runtime in the process not meaning CoreCLR in the process, would this follow similar semantics? i.e. if `hostfxr_initialize_for_runtime_config` is called twice (with compatible configs), only the context from the first call would allow setting properties?

Does compatibility of runtime configs mean that all properties need to match? e.g. if I initialize successfully with one config, then try to initialize with another config that has a property that wasn't defined in the host context with the first config, is that a failure to initialize?

vitek-karas replied:
Yes, only the first initialized host context will allow to set property values. I changed the wording and added some explanation. In theory there's a small window where the first host context is initialized but runtime has not started yet, where it could be possible to change properties even from other host contexts, but it's not worth the complexity. Whatever mechanism we choose to communicate which context was the first will have to be thread safe, so the native host should base its behavior on that flag.

Additional runtime configs specifying properties after the first context has been initialize - that is a problem. I'm not sure what to do here.
* On one hand it would be safer to simply fail - either on any property, or if the property has different value (comparing the values is a bit tricky, we would have to do case sensitive string match, but for example file paths still might be equal even if we don't recognize them as such...). There's also the added complexity of defaults, where if the property is not set, it typically has some kind of default value, but it can be argued that in that case the runtime config "doesn't care".
* On the other hand I can easily imagine cases where people using 3rd party components will want to load them regardless of runtime config properties. For example if the component for some super weird reason says that it wants server GC, it will very likely work just fine on workstation GC, but strict property comparison would fail.

Just thinking out loud:
* Maybe we could "ignore but notify"  - somehow keep the properties on the second context and return a special error code. It would be still possible to use the context to "run", with the expectation that the property values are ignored. 
* At the same time though we should somehow "merge" the properties from the first context.
* Maybe if the properties match - merge them so that in the "success" case the native host can inspect properties even on the second host context and get a correct and full picture of what the runtime sees.
* And if they don't match - return the special error code and don't merge, in which case the host would only see the properties coming from the second context - and thus be able to decide if it wants to "ignore" them and run anyway or not.
* Would be relatively complex behavior, but it might be the best of both worlds...

elinor-fung replied:
The ignore/notify and give the host a chance to decide behaviour sounds reasonable to me. I'd agree that it would be the best of both worlds.

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,209 @@
+
+*TODO: Should there be a way to enumerate all the properties?*
+
+*TODO: Currently it's possible to have duplicate properties, hosting components take properties from `.runtimeconfig.json` and then append the calculated ones without any conflict resolution, potentially creating duplicates. It's important to not that if there are duplicates, CoreCLR will fail to start very early on.*
```
~not~ -> note

I can't seem to figure out how to reply to an outdated comment, so from the other thread...
>Current runtime behavior is to fail early during startup if there are duplicates, so it's a situation which should really not happen, but we should have a defined behavior.

Can we make the 'initialize' fail and return a code indicating there are duplicates (since it would fail early during startup anyways)? Even though we don't resolve, we should be able to detect duplicates, right? Like you mentioned elsewhere, we're not aiming to allow conflict resolution between `.runtimeconfig.json` and the hosting components right now, it seems reasonable to me to fail before we get to allowing inspection/modification. (And if at some future point that is desired for some reason, we should be able to add an API for the host to get/resolve duplicates without breaking existing things that bail on initialize returning the error code for duplicates?)

vitek-karas replied:
That's actually a great idea! I think it's an acceptable change to fail on duplicates in the `hostpolicy` and don't wait on the runtime. The only observable behavior difference would be different failures
* Today depending on which property is duplicated and with which values the runtime may fail with different errors, but in the end it will always fail somehow.
* In the new world we would fail consistently the same way.

Let's actually do that regardless of the outcome of this document - it is a technically breaking change, but such a minor difference that we should be able to take it. And it makes the product much more predictable. And as you mention it let's us improve this in the future.

vitek-karas replied:
I filed dotnet/core-setup#5529 for this.

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,209 @@
+It needs to be possible to call this function simultaneously from multiple threads at the same time as well.
+It also needs to be possible to call this function while there is an active host context created by `hostfxr_initialize_for_app` and running inside the `hostfxr_run_app`.
+
+*TODO: What is the indication that there is no runtime in the process yet?*
```
Do we generally do different 'success' codes here?
I'm thinking of the S_OK / S_FALSE type thing - e.g. S_OK / 0 = was not previously initialized - initialized successfully, S_FALSE / 1 = was previously initialized - compatible with existing.

vitek-karas replied:
Yes - that's a good idea. Added that.

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,209 @@
+
+Setting properties is only supported when the host context was initialized without any runtime in the process. This is really a limitation of the runtime for which the runtime properties are immutable.
+
+*TODO: Can we support getting properties when there is preexisting runtime?*
```
Is there a reason to block this?

vitek-karas replied:
It's related to the discussion above - it's a bit tricky :-). I'll leave the TODO for now, until we figure out what we want to do here.

vitek-karas replied:
I removed this as we decided what to do with properties. I added a section about it into the doc.

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,209 @@
+* `type` - the type of runtime functionality requested - *TODO - exact names*
+  * `load_assembly` - entry point which loads an assembly (with dependencies) and returns function pointer for a specified static method.
+  * `com` - COM activation entry-point - see [COM activation](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/COM-activation.md) for more details.
+  * `ijw` - IJW entry-point - see [IJW activation](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/COM-activation.md) for more details.
```
```suggestion
  * `ijw` - IJW entry-point - see [IJW activation](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/IJW-activation.md) for more details.
```

vitek-karas replied:
Duh... :-(

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,209 @@
+  * `ijw` - IJW entry-point - see [IJW activation](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/COM-activation.md) for more details.
+* `delegate` - when successful, the native function pointer to the requested runtime functionality.
+
+Initially the function will only work if `hostfxr_initialize_for_runtime_config` was used to initialize the host context. Later on this could be relaxed to allow the combination with `hostfxr_initialize_for_app`.  
```
>allow the combination

allow being used in combination

vitek-karas committed:
PR feedback

AaronRobinsonMSFT commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,213 @@
+## New `hostfxr` APIs
+
+Terminology:
+* "native host" - the code which uses the proposed APIs. Typically a native (C++) application, but can really be any "non .NET Core" code.
```
>Typically a native (C++) application

In the spirit of precision this is a bad idea. Although we can use C++, all exposed function contracts must be marked `extern "C"` or specifically have C-style linking. Otherwise we are going to be exposed to compiler mangling sadness. Ideally we should define a "native host" more closely to the latter part of the sentence "can be any "non .NET Core" code." Alternatively or additionally we can specify that a "native host" is any application that explicitly activates CoreCLR through the hosting C-style API.

vitek-karas replied:
Changed the wording to avoid specific languages and such.

AaronRobinsonMSFT commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,213 @@
+* Process the `.runtimeconfig.json`
+* Resolve framework references and find actual frameworks
+* Find the root framework (`Microsoft.NETCore.App`) and load the `hostpolicy` from it
+* The `hostpolicy` will then process all relevant `.deps.json` files and produce the TPA, native search paths and other artifacts.
```
There are so many acronyms in the hosting world. Can we define "TPA" and "TFM" or at least provide a link to a doc with the definitions.

vitek-karas replied:
Removed the acronym...

AaronRobinsonMSFT commented on Documentation/design-docs/native-hosting-initialize.md:
````diff
@@ -0,0 +1,213 @@
+* `dotnet_root` - path to the root of the .NET Core installation in use. This typically points to the install location from which the `hostfxr` has been loaded. For example on Windows this would typically point to `C:\Program Files\dotnet`. The path is used to search for shared frameworks and potentially SDKs.
+
+
+``` C++
````
This is an export right? If so please update the language to be C or prefix the signature with `extern "C"`. It also might be a good place to comment on calling convention or link to the ABI contracts the hosting API follows.

vitek-karas replied:
I was trying to avoid to write the full blown header file in the spec... I'll change the language to C, but I think including full calling convention specification in the spec feels abit over the top. (I agree it's important, but not in the spec I think)

vitek-karas replied:
Added a note about exports and that they follow the same style as the other exports in `hostfxr`.

AaronRobinsonMSFT commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,213 @@
+
+Initializes the hosting components for running a managed application.
+When used to execute an app, the `app_path` (or CLI equivalent) will be used to locate the `.runtimeconfig.json` and the `.deps.json` which will be used to load the application and its dependent frameworks.
+* `argc` and `argv` - the command line - entirely optional
```
Is the optional value of `argc` `0` or `-1`? If `argc` is `0` does that mean that `argv` can be `nullptr`? Instead of specifying this we could simply state if `argc` is `0`, then `argv` is ignored?

AaronRobinsonMSFT commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,213 @@
+Initializes the hosting components for running a managed application.
+When used to execute an app, the `app_path` (or CLI equivalent) will be used to locate the `.runtimeconfig.json` and the `.deps.json` which will be used to load the application and its dependent frameworks.
+* `argc` and `argv` - the command line - entirely optional
+* `app_path` - path to the application (the managed `.dll`) to run. This can be `nullptr` if the app is specified in the command line arguments.
```
>This can be `nullptr` if the app is specified in the command line arguments.

Do we really want this behavior? It reminds me of Win32's [`CreateProcess()`](https://docs.microsoft.com/windows/desktop/api/processthreadsapi/nf-processthreadsapi-createprocessa) and the constant confusion between the `lpApplicationName` and `lpCommandLine` parameters. Take a few and read the documentation on those - they are amazing.

vitek-karas replied:
Well - this is actually describing the current behavior of the `hostfxr` APIs. It comes from the desire to be able to use these APIs for all scenarios. We can limit this I guess.. to make the API cleaner.

vitek-karas replied:
In a way we should support the case where everything comes from the command line - if I would want to write a host which runs apps (basically a second `dotnet.exe`) that's exactly what would be needed, since the command line can contain host specific options before the app name. For example `dotnet.exe --roll-forward Major app.dll`. So asking the native host code to be able to parse this correct would be tricky.

But maybe it's not a scenario just yet (I'm not entire sure about ASP.NET, it might be required there).

AaronRobinsonMSFT replied:
I am fine with it, if those are the existing semantics. I just wouldn't want to create a new API with that kind of behavior.

vitek-karas replied:
I don't think we need the super specific description like `CreateProcess`. In our case it could be as simple as:
* non-null - this application will be executed
* null - the argv/argc specify the command in the format supported by `dotnet.exe` (doc link). Maybe mentioning that we would not want to support SDK commands this way.

We might do this as two APIs, but then those are very similar.

AaronRobinsonMSFT commented on Documentation/design-docs/native-hosting-initialize.md:
````diff
@@ -0,0 +1,213 @@
+);
+```
+
+This function would load the specified `.runtimeconfig.json`, resolve all frameworks, resolve all the assets from those frameworks abd then prepare runtime initialization where the TPA contains only frameworks. Note that this case does NOT consume any `.deps.json` from the app/component (only processes the framework's `.deps.json`). This entry point is intended for `comhost`/`ijwhost`/`nethost` a ...
````
Instead of consuming a path, I would suggest we consume the read in configuration file. This would allow embedding the configuration as well as allow us to avoid assumptions about how to access the environment's file system. Is there a reason we want to always read from disk? Might be interesting for ASP.NET or other scenarios to pipe in config from somewhere and with this API they would have to write to disk. Instead they could just pass a buffer with the configuration contents?

vitek-karas replied:
Depending on discussions around other APIs, this might be the main way for component hosting (the easy to use API seems to be hard to specify... we may not be able to get one in time). In such case having to load the file into memory seems like rather unnecessary burden.

The thinking was that we could add another version which loads it from memory. There's also some desire to be able to specify it as a structured parameter - not JSON, some kind of struct.

vitek-karas committed:
PR feedback

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,215 @@
+* Process the `.runtimeconfig.json`
+* Resolve framework references and find actual frameworks
+* Find the root framework (`Microsoft.NETCore.App`) and load the `hostpolicy` from it
+* The `hostpolicy` will then process all relevant `.deps.json` files and produce the list of assemblies, native search paths and other artifacts needed to initialize th
```
...e runtime?

vitek-karas replied:
Duh - no idea what happened.

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,215 @@
+
+This function will fail if there already is a CoreCLR running in the process.
+
+*Note: This is effectively a replacement for `hostfxr_main_startupinfo` and `hostfxr_main`. Currently it is not be a goal to fully replace these APIs because they also support SDK commands which are special in lot of ways and don't fit well with the rest of the native hosting. There's no scenario right now which would require the ability to issue SDK commands from a native host. That said nothing ...
```
is not ~be~ a goal

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
````diff
@@ -0,0 +1,215 @@
+```
+
+This function would load the specified `.runtimeconfig.json`, resolve all frameworks, resolve all the assets from those frameworks abd then prepare runtime initialization where the TPA contains only frameworks. Note that this case does NOT consume any `.deps.json` from the app/component (only processes the framework's `.deps.json`). This entry point is intended for `comhost`/`ijwhost`/`nethost` a ...
+* `runtime_config_path` - path to the `.runtimeconfig.json` file to process. Unlike with the `hostfxr_initialize_for_app`, there is no `.deps.json` from the app/component which will be processed.
````
>there is no `.deps.json` from the app/component which will be processed

any `.deps.json` from the app/component will not be processed

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,215 @@
+
+Trying to get a property which doesn't exist is an error and `hostfxr_get_runtime_property` will return an appropriate error code.
+
+Setting properties is only supported on the first host context in the process. This is really a limitation of the runtime for which the runtime properties are immutable. Once the first host context is initialized and starts a runtime there's no way to change these properties. For now we will not consider the scenario where the host context is initialize but the runtime hasn't started yet, mainly  ...
```
```suggestion
Setting properties is only supported on the first host context in the process. This is really a limitation of the runtime for which the runtime properties are immutable. Once the first host context is initialized and starts a runtime there's no way to change these properties. For now we will not consider the scenario where the host context is initialized but the runtime hasn't started yet, mainly for simplicity of implementation and lack of requirements.
```

vitek-karas committed:
Describe runtime property resolution for secondary host contexts

PR feedback

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
````diff
@@ -0,0 +1,221 @@
+);
+```
+
+This function would load the specified `.runtimeconfig.json`, resolve all frameworks, resolve all the assets from those frameworks abd then prepare runtime initialization where the TPA contains only frameworks. Note that this case does NOT consume any `.deps.json` from the app/component (only processes the framework's `.deps.json`). This entry point is intended for `comhost`/`ijwhost`/`nethost` a ...
````
```suggestion
This function would load the specified `.runtimeconfig.json`, resolve all frameworks, resolve all the assets from those frameworks and then prepare runtime initialization where the TPA contains only frameworks. Note that this case does NOT consume any `.deps.json` from the app/component (only processes the framework's `.deps.json`). This entry point is intended for `comhost`/`ijwhost`/`nethost` and similar scenarios.
```

elinor-fung commented on Documentation/design-docs/native-hosting-initialize.md:
```diff
@@ -0,0 +1,221 @@
+* The initialization will compare the runtime properties from the `.runtimeconfig.json` specified in the `runtime_config_path` with those already set to the runtime in the process
+  * If all properties from the new runtime config are already set and have the exact same values (case sensitive string comparison), the initialization succeeds with no additional consequences. (Note that this is the most typical case where the runtime config have no properties in it.)
+  * If there are either new properties which are not set in the runtime or ones which have different values, the initialization will return a special return code - a "warning". It's not a full on failure as initialized context will be returned.
+  * In both cases only the properties specified by the new runtime config will be reported on the host context. This is to allow host to decide in the "warning" case if it's OK to let the component run or not.
```
```suggestion
  * In both cases only the properties specified by the new runtime config will be reported on the host context. This is to allow the native host to decide in the "warning" case if it's OK to let the component run or not.
```

elinor-fung replied:
How will the native host know which properties of the new config were in conflict / ignored in order to make the decision? Return code on `hostfxr_get_runtime_property`?

vitek-karas replied:
My current thinking is that we let the native host access properties on the "current runtime" - via passing `nullptr` as the host context to `hostfxr_get_runtime_property`. And then the native host can compare by itself to figure out which one was "wrong". I don't think we need to make it easy, just possible.

vitek-karas replied:
I added the ability to use `nullptr` context into the document.

elinor-fung replied:
Sounds good!

vitek-karas committed:
Add ability to enumerate runtime properties

Properties can now be enumerated.
Also properties on the first context are accessible via the null handle.
Reorganized the properties section in the document to make it more readable.

PR feedback
